File: python/nerf_model.py
import numpy as np
import torch
import torch.nn as nn
from radiance_field import RadianceField
from nerf_function import split_ray, sample_coarse, sample_fine, rgb_and_weight
from typing import Tuple


class NeRF(nn.Module):

    # sampling parameter
    N_c = 64
    N_f = 128

    BATCH_SIZE = 2048

    def __init__(self, t_n: float = 0., t_f: float = 2.5, L_x: int = 10, L_d: int = 4) -> None:
        self.t_n = t_n
        self.t_f = t_f
        self.c_bg = (1, 1, 1)

        super(NeRF, self).__init__()
        self.rf_c = RadianceField(L_x=L_x, L_d=L_d)
        self.rf_f = RadianceField(L_x=L_x, L_d=L_d)

    def device(self) -> torch.device:
        return next(self.parameters()).device

    def forward(self, o: torch.Tensor, d: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]:
        device = self.device()
        o = o.to(device)
        d = d.to(device)

        batch_size = o.shape[0]
        device = o.device

        partitions = split_ray(self.t_n, self.t_f, self.N_c, batch_size)
        partitions = partitions.to(device)
        partitions = partitions.detach()

        # background.
        bg = torch.tensor(self.c_bg, device=device, dtype=torch.float32)
        bg = bg.view(1, 3)

        # coarse rendering.
        _t_c = sample_coarse(partitions)
        t_c = _t_c.to(device)

        rgb_c, w_c = rgb_and_weight(self.rf_c, o, d, t_c, self.N_c)
        C_c = torch.sum(w_c[..., None] * rgb_c, dim=1)
        C_c += (1. - torch.sum(w_c, dim=1, keepdim=True)) * bg

        # fine rendering.
        _w_c = w_c.clone()
        t_f = sample_fine(partitions, _w_c, _t_c, self.N_f)
        t_f = t_f.to(device)

        rgb_f, w_f = rgb_and_weight(self.rf_f, o, d, t_f, self.N_f + self.N_c)
        C_f = torch.sum(w_f[..., None] * rgb_f, dim=1)
        C_f += (1. - torch.sum(w_f, dim=1, keepdim=True)) * bg

        return C_c, C_f

File: cpp/nerf_model.cpp
#include "nerf_model.hpp"

#include "nerf_function.hpp"

NeRFImpl::NeRFImpl(float _t_n, float _t_f, int32_t _L_x, int32_t _L_d, cv::Vec3b _c_bg)
    : t_near_(_t_n), t_far_(_t_f), back_ground_color_(_c_bg) {
  using namespace torch::nn;
  rf_coarse_ = register_module("rf_coarse_", RadianceField(_L_x, _L_d));
  rf_fine_ = register_module("rf_fine_", RadianceField(_L_x, _L_d));
}

torch::Device NeRFImpl::device() { return rf_coarse_->parameters().front().device(); }

std::pair<torch::Tensor, torch::Tensor> NeRFImpl::forward(torch::Tensor o, torch::Tensor d) {
  const torch::Device dev = device();
  o = o.to(dev);
  d = d.to(dev);

  const int32_t batch_size = o.size(0);

  torch::Tensor partitions = SplitRay(t_near_, t_far_, kNumCoarse, batch_size);
  partitions = partitions.to(dev);

  std::vector<float> bg_vec(back_ground_color_.val, back_ground_color_.val + 3);
  torch::Tensor bg = torch::tensor(bg_vec, torch::TensorOptions().device(dev).dtype(torch::kFloat32));
  bg = bg.view({1, 3});
  bg /= 255;

  torch::Tensor t_c = SampleCoarse(partitions).to(dev);
  t_c = t_c.to(dev);

  auto [rgb_c, w_c] = _rgb_and_weight(rf_coarse_, o, d, t_c, kNumCoarse);
  torch::Tensor C_c = torch::sum(w_c.unsqueeze(-1) * rgb_c, 1);
  C_c += (1.0f - torch::sum(w_c, 1, true)) * bg;

  w_c = w_c.clone();
  torch::Tensor t_f = SampleFine(partitions, w_c, t_c, kNumFine).to(dev);

  auto [rgb_f, w_f] = _rgb_and_weight(rf_fine_, o, d, t_f, kNumFine + kNumCoarse);
  torch::Tensor C_f = torch::sum(w_f.unsqueeze(-1) * rgb_f, 1);
  C_f += (1.0f - torch::sum(w_f, 1, true)) * bg;

  return std::make_pair(C_c, C_f);
}
